x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java

```diff
@@ -754,7 +756,7 @@ private void runComputeOnDataNode(
         } catch (Exception e) {
             exchangeService.finishSinkHandler(externalId, e);
             exchangeService.finishSinkHandler(request.sessionId(), e);
-            listener.onFailure(e);
+            listenerRefs.acquire().onFailure(e);
```

Comment:
This is the main change - we should notify the sub-listener, not the main listener.